File: core/src/com/bombbird/terminalcontrol/entities/sidstar/Sid.java
package com.bombbird.terminalcontrol.entities.sidstar;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.entities.airports.Airport;
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;

public class Sid extends SidStar {
    private HashMap<String, int[]> initClimb;
    private HashMap<String, Array<Waypoint>> initWpts;
    private HashMap<String, Array<int[]>> initRestrictions;
    private HashMap<String, Array<Boolean>> initFlyOver;
    private Array<Array<String>> transition;
    private String[] centre;

    public Sid(Airport airport, JSONObject jo) {
        super(airport, jo);
    }

    public Sid(Airport airport, JSONArray wpts, JSONArray restrictions, JSONArray fo, String name) {
        super(airport, wpts, restrictions, fo, name);

        initClimb = new HashMap<>();
        initWpts = new HashMap<>();
        initRestrictions = new HashMap<>();
        initFlyOver = new HashMap<>();
    }

    @Override
    public void parseInfo(JSONObject jo) {
        super.parseInfo(jo);

        initClimb = new HashMap<>();
        initWpts = new HashMap<>();
        initRestrictions = new HashMap<>();
        initFlyOver = new HashMap<>();
        transition = new Array<>();

        JSONObject rwys = jo.getJSONObject("rwys");
        for (String rwy: rwys.keySet()) {
            getRunways().add(rwy);
            Array<Waypoint> wpts = new Array<>();
            Array<int[]> restrictions = new Array<>();
            Array<Boolean> flyOver = new Array<>();

            JSONObject rwyObject = rwys.getJSONObject(rwy);
            String[] initClimbData = rwyObject.getString("climb").split(" ");
            JSONArray initWptData = rwyObject.getJSONArray("wpts");
            for (int i = 0; i < initWptData.length(); i++) {
                String[] data = initWptData.getString(i).split(" ");
                String wptName = data[0];
                wpts.add(TerminalControl.radarScreen.waypoints.get(wptName));
                restrictions.add(new int[] {Integer.parseInt(data[1]), Integer.parseInt(data[2]), Integer.parseInt(data[3])});
                boolean fo = data.length > 4 && data[4].equals("FO");
                flyOver.add(fo);
                if (fo) Waypoint.flyOverPts.put(wptName, true);
            }
            initClimb.put(rwy, new int[] {Integer.parseInt(initClimbData[0]), Integer.parseInt(initClimbData[1]), Integer.parseInt(initClimbData[2])});
            initWpts.put(rwy, wpts);
            initRestrictions.put(rwy, restrictions);
            initFlyOver.put(rwy, flyOver);
        }

        JSONArray transitions = jo.getJSONArray("transitions");
        for (int i = 0; i < transitions.length(); i++) {
            JSONArray trans = transitions.getJSONArray(i);
            Array<String> transData = new Array<>();
            for (int j = 0; j < trans.length(); j++) {
                transData.add(trans.getString(j));
            }
            transition.add(transData);
        }

        JSONArray control = jo.getJSONArray("control");
        centre = new String[2];
        centre[0] = control.getString(0);
        centre[1] = control.getString(1);
    }

    public Array<String> getRandomTransition() {
        return transition.get(MathUtils.random(0, transition.size - 1));
    }

    public int[] getInitClimb(String rwy) {
        return initClimb.get(rwy);
    }

    public String[] getCentre() {
        return centre;
    }

    public Array<Waypoint> getInitWpts(String runway) {
        return initWpts.get(runway);
    }

    public Array<int[]> getInitRestrictions(String runway) {
        return initRestrictions.get(runway);
    }

    public Array<Array<String>> getTransition() {
        return transition;
    }

    public Array<Boolean> getInitFlyOver(String runway) {
        return initFlyOver.get(runway);
    }
}

File: core/src/com/bombbird/terminalcontrol/entities/sidstar/Sid.kt
package com.bombbird.terminalcontrol.entities.sidstar

import com.bombbird.terminalcontrol.entities.airports.Airport
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint
import com.badlogic.gdx.utils.Array
import org.json.JSONArray
import org.json.JSONObject
import kotlin.collections.HashMap

class Sid : SidStar {
    private val initClimb = HashMap<String, IntArray>()
    private val initRouteData = HashMap<String, RouteData>()
    private var transition = Array<Array<String>>()
    var centre = arrayOf("Control", "125.5")
        private set

    constructor(airport: Airport, jo: JSONObject) : super(airport) {
        parseInfo(jo)
    }

    constructor(airport: Airport, wpts: JSONArray, restrictions: JSONArray, fo: JSONArray, name: String) : super(airport, wpts, restrictions, fo, name)

    override fun parseInfo(jo: JSONObject) {
        super.parseInfo(jo)
        val rwys = jo.getJSONObject("rwys")
        for (rwy in rwys.keySet()) {
            runways.add(rwy)
            val routeData = RouteData()
            val rwyObject = rwys.getJSONObject(rwy)
            val initClimbData = rwyObject.getString("climb").split(" ".toRegex()).toTypedArray()
            val initWptData = rwyObject.getJSONArray("wpts")
            for (i in 0 until initWptData.length()) {
                val data = initWptData.getString(i).split(" ".toRegex()).toTypedArray()
                val wptName = data[0]
                routeData.add(
                    radarScreen.waypoints[wptName]!!,
                    intArrayOf(data[1].toInt(), data[2].toInt(), data[3].toInt()),
                    data.size > 4 && data[4] == "FO"
                )
            }
            initClimb[rwy] = intArrayOf(initClimbData[0].toInt(), initClimbData[1].toInt(), initClimbData[2].toInt())
            initRouteData[rwy] = routeData
        }
        val transitions = jo.getJSONArray("transitions")
        for (i in 0 until transitions.length()) {
            val trans = transitions.getJSONArray(i)
            val transData = Array<String>()
            for (j in 0 until trans.length()) {
                transData.add(trans.getString(j))
            }
            transition.add(transData)
        }
        val control = jo.getJSONArray("control")
        centre[0] = control.getString(0)
        centre[1] = control.getString(1)
    }

    val randomTransition: Array<String>?
        get() = if (transition.isEmpty) null else transition.random()

    fun getInitClimb(runway: String?): IntArray? {
        if (runway == null) return null
        return initClimb[runway]
    }

    fun getInitWpts(runway: String?): Array<Waypoint>? {
        if (runway == null) return null
        return initRouteData[runway]?.waypoints
    }

    fun getInitRestrictions(runway: String?): Array<IntArray>? {
        if (runway == null) return null
        return initRouteData[runway]?.restrictions
    }

    fun getInitFlyOver(runway: String?): Array<Boolean>? {
        if (runway == null) return null
        return initRouteData[runway]?.flyOver
    }

    fun isRadarDep(runway: String?): Boolean {
        if (runway == null) return false
        return initRouteData[runway]?.size == 0 && routeData.size == 0
    }
}